lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -50,44 +55,121 @@ links offset = level_offsets[level] + offsets[reindex[point_id]]
 
 const HEADER_SIZE: usize = 64;
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 struct GraphLinksFileInfo {
     point_count: usize,
     reindex_start: usize,
     links_start: usize,
     offsets_start: usize,
     offsets_end: usize,
+    compression: Option<CompressionInfo>,
+}
+
+#[derive(Clone, Debug)]
+struct CompressionInfo {
+    m: usize,
+    m0: usize,
+    bits_per_unsorted: u8,
 }
 
 #[derive(AsBytes, FromBytes, FromZeroes)]
 #[repr(C)]
-struct GraphLinksFileHeader {
+struct HeaderV0 {
     point_count: u64,
     levels_count: u64,
-    total_links_len: u64,
-    total_offsets_len: u64,
-    offsets_padding: u64, // 0 or 4
+    total_links_count: u64,
+    total_offset_count: u64,
+    /// Either 0 or 4.
+    offsets_padding_bytes: u64,
+    zero_padding: [u8; 24],
 }
 
+/// File header for the compressed format.
+#[derive(AsBytes, FromBytes, FromZeroes)]
+#[repr(C)]
+struct HeaderV1 {
+    point_count: U64,
+    /// Should be [`HEADER_VERSION_1`].
+    ///
+    /// Deliberately placed at the same offset as [`HeaderV0::levels_count`] and
+    /// set to an impossibly large number to make old Qdrant versions fail fast
+    /// when trying to read the new format.
+    version: U64,
+    levels_count: U64,
+    total_links_bytes: U64,
+    total_offset_count: U64,
+    m: U64,
+    m0: U64,
+    zero_padding: [u8; 8],
+}
+
+const HEADER_VERSION_1: u64 = 0xFFFF_FFFF_FFFF_FF01;
```

Comment:
I prefer to have `0xFFFF_FFFF_FFFF_FF01` explicitly written in the code since it's the value that is visible in `hexdump graph.links | head -1`. More convenient for debugging/troubleshooting.